src/core/markovProcessGraph.js

```diff
@@ -0,0 +1,579 @@
+// @flow
+
+/**
+ * Data structure representing a particular kind of Markov process, as
+ * kind of a middle ground between the semantic SourceCred graph (in the
+ * `core/graph` module) and a literal transition matrix. Unlike the core
+ * graph, edges in a Markov process graph are unidirectional, edge
+ * weights are raw transition probabilities (which must sum to 1) rather
+ * than unnormalized weights, and there are no dangling edges. Unlike a
+ * fully general transition matrix, parallel edges are still reified,
+ * not collapsed; nodes have weights, representing sources of flow, and
+ * a few SourceCred-specific concepts are made first-class:
+ * specifically, cred minting and time period fibration. The
+ * "teleportation vector" from PageRank is also made explicit via the
+ * "adjoined seed node" graph transformation strategy, so this data
+ * structure can form well-defined Markov processes even from graphs
+ * with nodes with no out-weight. Because the graph reifies the
+ * teleportation and temporal fibration, the associated parameters are
+ * "baked in" to weights of the Markov process graph.
+ *
+ * We use the term "fibration" to refer to a graph transformation where
+ * each scoring node is split into one node per epoch, and incident
+ * edges are rewritten to point to the appropriate epoch nodes. The term
+ * is vaguely inspired from the notion of a fiber bundle, though the
+ * analogy is not precise.
+ *
+ * The Markov process graphs in this module have three kinds of nodes:
+ *
+ *   - *contribution nodes*, which are in 1-to-1 correspondence with the
+ *     nodes in the underlying core graph (including users);
+ *   - *epoch nodes*, which are created for each time period for each
+ *     scoring node; and
+ *   - the *seed node*, which reifies the teleportation vector and
+ *     forces well-definedness and ergodicity of the Markov process (for
+ *     nonzero alpha, and assuming that there is at least one edge in
+ *     the underlying graph).
+ *
+ * The edges include:
+ *
+ *   - *contribution edges* between nodes in the underlying graph, which
+ *     are lifted to their corresponding contribution nodes or to epoch
+ *     nodes if either endpoint has been fibrated;
+ *   - *seed-in* / "redistribution" / "radiation" edges from nodes to
```

Comment:
> radiation

This name occurred to me while mulling. The implication is twofold[.][1]
First, the seed-in edges cause the normal cred flow to “lose energy” due
to a background radiation of warmth back to the root. But also seed-in
edges are required for the Markov process to be well defined, because
they define the action at “black hole” nodes—and black holes dissipate
via Hawking radiation.

I’d like *some* name better than “seed-in”, which I feel isn’t entirely
unambiguous (node *v*’s “seed-in” edge is actually an out-edge of *v*).
*Redistribution* is more direct, but sometimes an appropriate splash of
color doesn’t hurt, either. Thoughts welcome. :-)

[1]: https://en.wikipedia.org/wiki/%C4%AAh%C4%81m